### PR TITLE
 Application specific override for GPU scheduling behavior

### PR DESCRIPTION
**Pull request by meln1k**
https://github.com/mesosphere/marathon/pull/6051 needs to be merged first!

 Application specific override for GPU scheduling behavior

Summary: Implemented an application-specific GPU scheduling behavior override

JIRA issues: MARATHON-8089

### CODE REVIEW

File: src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala [hunk 1]
```diff
@@ -1179,6 +1179,60 @@ class ResourceMatcherTest extends UnitTest with Inside {
...
+        "--draining_seconds", "300",
+        "--gpu_scheduling_behavior", "unrestricted",
+        "--enable_features", "gpu_resources")
+      val offer = MarathonTestHelper.makeBasicOffer(gpus = 4)
+        .build()
+      val app = AppDefinition(
+        id = "/test".toRootPath,
+        resources = Resources(cpus = 1.0, mem = 128.0, disk = 0.0, gpus = 0),
+        portDefinitions = PortDefinitions(0, 0),
+        labels = Map("GPU_SCHEDULING_BEHAVIOR" -> "restricted")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala]: I think we are missing tests for no label. What do you think about using a property table?

```scala
Table(
    ("gpu_scheduling_behavior", "GPU_SCHEDULING_BEHAVIOR", expected),
    ("unrestricted", Some("restricted"), Match),
    ("unrestricted", None, Match),
    ("restricted", Some("unrestricted"), NoMatch),
    ("restricted", None, NoMatch),
)
```

This seems easier to comprehend to me.

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 2]
```diff
@@ -249,9 +249,11 @@ object ResourceMatcher extends StrictLogging {
     }
 
     val checkGpuSchedulingBehaviour: Boolean = {
+      val applicationSpecificGpuBehavior = runSpec.labels.get("GPU_SCHEDULING_BEHAVIOR")
+        .filter(behavior => Set(GpuSchedulingBehavior.Restricted, GpuSchedulingBehavior.Unrestricted).contains(behavior))
```
Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: do we want to create this set for every check? Maybe we should create that on the ResourceMatcher level at least?

Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: do we want to create this set for every check? Maybe we should create that on the ResourceMatcher level at least?

File: src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala [hunk 3]
```diff
@@ -1179,6 +1180,46 @@ class ResourceMatcherTest extends UnitTest with Inside {
...
+          "--gpu_scheduling_behavior", gpuSchedulingBehavior,
+          "--enable_features", "gpu_resources")
+        val offer = MarathonTestHelper.makeBasicOffer(gpus = 4)
+          .build()
+        val app = AppDefinition(
+          id = "/test".toRootPath,
+          resources = Resources(cpus = 1.0, mem = 128.0, disk = 0.0),
+          portDefinitions = PortDefinitions(0, 0),
+
+          labels = overrideLabel.map(label => Map("GPU_SCHEDULING_BEHAVIOR" -> label)).getOrElse(Map.empty)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala]: Nit: Is `overrideLabel.map( ... Map)` not returning a `Option[Map]`?

File: src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala [hunk 4]
```diff
@@ -1179,6 +1180,46 @@ class ResourceMatcherTest extends UnitTest with Inside {
...
+
+      val overrideCases = Table(
+        ("gpu_scheduling_behavior", "GPU_SCHEDULING_BEHAVIOR", "expected"),
+        ("unrestricted", Some("restricted"), "NoMatch"),
+        ("unrestricted", None, "Match"),
+        ("restricted", Some("unrestricted"), "Match"),
+        ("restricted", None, "NoMatch")
+      )
+
+      forAll(overrideCases) { (gpuSchedulingBehavior, overrideLabel, expected) =>
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala]: Nit: I'd like to use these cases outside of the `.. should ... in` pattern so we create new cases. It's not a blocker though.

File: src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala (hunk 3 above)
Comment (meln1k) [reply src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala]: it does!

Comment (jeschkies) [reply src/test/scala/mesosphere/mesos/ResourceMatcherTest.scala]: Oh, I missed the `getOeElse` part.

**Pull request merged**